core/src/main/scala/com.snowplowanalytics.stream/loader/clients/BulkSender.scala

```diff
@@ -122,7 +122,7 @@ object BulkSender {
     }
   }
 
-  def futureToTask[T](f: => Future[T]): IO[T] =
+  def futureToTask[T](f: => Future[T])(implicit cs: ContextShift[IO]): IO[T] =
     IO.fromFuture(IO.delay(f))
```

Comment:
Given that the app is fully synchronous, I start to wonder if there is any benefit in introducing `IO` at all. Do you see a reason why we should not drop it?

Reply:
It's because the elastic4s client returns a Future, so we have to do _something_ with the Future.  We have two choices:

1. Await the future directly
2. Convert it to IO and then wait for the result of the IO.

We do the latter, and I think it's because then we use the nice cats-retry syntax.  That's the only reason I can think of.

I can imagine an interesting rewrite of this code like this:

- Drop the elastic4s wrapper entirely
- Use the elasticsearch java client directly.
- Call the synchronous method [performRequest](https://github.com/elastic/elasticsearch/blob/main/client/rest/src/main/java/org/elasticsearch/client/RestClient.java#L286) to get the response back on the same thread.

I didn't do it because I was trying to make these fixes as small as possible, without changing too much.

Reply:
I'd be in favor of dropping `elastic4s`, but as a quicker solution I'd be in favor of

```
1. Await the future directly
```

so that we remove a (quite useless) layer. We can implement the retries without `IO`.

But I'm fine with the current implementation with the timeout on the `IO`.